app/Http/Controllers/ListController.php

```diff
@@ -2,131 +2,126 @@
-use Illuminate\Database\Query\Builder;
 use Illuminate\Http\Request;
 use Illuminate\Http\Response;
+use Illuminate\Support\Collection;
 use Illuminate\Support\Facades\DB;
 use Laravel\Lumen\Routing\Controller;
+use Illuminate\Database\Query\Builder;
 use Symfony\Component\HttpKernel\Exception\NotFoundHttpException;
-use Symfony\Component\HttpKernel\Exception\UnprocessableEntityHttpException;
 
 class ListController extends Controller
 {
-    public function get(Request $request, string $listId)
+    protected $tableName = 'materials';
+    protected $idColumn = 'material';
+    protected $idFilterName = 'material_ids';
+
+    public function get(Request $request, ListType $type): array
     {
-        $this->checkList($listId);
+        return [
+            'id' => $type,
+            'materials' => $this->getItems($request, $type),
+        ];
+    }
 
-        $query = DB::table('materials')
-            ->where(['guid' => $request->user()->getId(), 'list' => $listId]);
+    protected function getItems(Request $request, ListType $type): Collection
+    {
+        $query = DB::table($this->tableName)
+            ->where(['guid' => $request->user()->getId(), 'list' => $type]);
 
-        // Filter to the given materials, if supplied.
-        if ($request->has('material_ids')) {
+        // Filter to the given items, if supplied.
+        $itemIds = $this->commaSeparatedQueryParamToArray($this->idFilterName, $request);
+        if (count($itemIds)) {
             // The OpenAPI spec defines the parameter as a comma separated
             // list. OpenAPI defaults to using "id=1&id=2" for array types,
             // but PHP expects "id[]=1&id[]=2". So rather than trying to hack
             // around that, we use the other common option of using a single
             // comma separated value and just split it up here. Looks nicer in
             // the URL.
-            $ids = explode(',', $request->get('material_ids'));
-            $query->where(function ($query) use ($ids) {
-                foreach ($ids as $id) {
-                    $this->materialQuery($query, $id, true);
+            $query->where(function ($query) use ($itemIds) {
+                foreach ($itemIds as $itemId) {
+                    $item = ListItem::createFromUrlParameter($itemId);
+                    $this->idQuery($query, $item, true);
                 }
             });
         }
 
-        $materials = $query->orderBy('changed_at', 'DESC')
-            ->select('material')
-            ->pluck('material');
+        $items = $query->orderBy('changed_at', 'DESC')
+            ->select($this->idColumn)
+            ->pluck($this->idColumn);
 
-        return [
-            'id' => $listId,
-            'materials' => $materials,
-        ];
+        return $items;
     }
 
-    public function checkMaterial(Request $request, string $listId, string $materialId)
+    public function itemAvailability(Request $request, ListType $type, ListItem $item): Response
```

Comment:
Hmm, I thought about that name too, but I choose itemAvailability because I normally strictly use that kind of naming for functions that only returns boolean.
But what the heck hasItem it is - no strong feelings.

Reply:
Don't let dogma get in the way of clarity. :smiley: `itemOnList` could work, but then one kinda misses the being able to use questionmark in method names.

This isn't a class you'll be using directly in your own code, so I can defend the method not returning a bool.